samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java

```diff
@@ -0,0 +1,199 @@
+    metricsReg.newGauge(GROUP_NAME, new SupplierGauge(tableId + "-miss-rate", () -> getRate(missCount)));
+    metricsReg.newGauge(GROUP_NAME, new SupplierGauge(tableId + "-req-count", () -> requestCount()));
+  }
+
+  @Override
+  public V get(K key) {
+    V value = cache.get(key);
+    if (value == null) {
+      ++missCount;
+      value = atomicGet(key);
+    } else {
+      ++hitCount;
+    }
+    return value;
+  }
+
+  @Override
+  public Map<K, V> getAll(List<K> keys) {
+    Map<K, V> retMap = new HashMap<>();
+    keys.stream().forEach(k -> retMap.put(k, get(k)));
+    return retMap;
+  }
+
+  @Override
+  public void close() {
+    this.cache.close();
+    this.rdTable.close();
+  }
+
+  @Override
+  public void put(K key, V value) {
+    Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
+    atomicPut(key, value);
+  }
+
+  @Override
+  public void putAll(List<Entry<K, V>> entries) {
+    Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
+    entries.forEach(e -> put(e.getKey(), e.getValue()));
+  }
+
+  @Override
+  public void delete(K key) {
+    Preconditions.checkNotNull(rwTable, "Cannot delete from a read-only table: " + rdTable);
+    atomicDelete(key);
+  }
+
+  @Override
+  public void deleteAll(List<K> keys) {
+    Preconditions.checkNotNull(rwTable, "Cannot delete from a read-only table: " + rdTable);
+    keys.stream().forEach(k -> delete(k));
+  }
+
+  @Override
+  public synchronized void flush() {
+    Preconditions.checkNotNull(rwTable, "Cannot flush a read-only table: " + rdTable);
+    rwTable.flush();
+  }
+
+  private V atomicGet(K key) {
+    Lock lock = stripedLocks.get(key);
+    V value = null;
+    try {
+      lock.lock();
+      if (cache.get(key) == null) {
+        // Due to the lack of contains() API in ReadableTable, there is
+        // no way to tell whether a null return by cache.get(key) means
+        // cache miss or the value is actually null. As such, we cannot
+        // support negative cache semantics.
+        value = rdTable.get(key);
+        if (value != null) {
+          cache.put(key, value);
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+
+    return value;
+  }
+
+  private void atomicPut(K key, V value) {
+    Lock lock = stripedLocks.get(key);
+    try {
+      lock.lock();
+      rwTable.put(key, value);
+      if (!isWriteAround) {
+        cache.put(key, value);
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void atomicDelete(K key) {
+    Lock lock = stripedLocks.get(key);
+    try {
+      lock.lock();
+      rwTable.delete(key);
+      cache.delete(key);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private double getRate(long count) {
+    long reqs = requestCount();
+    return reqs == 0 ? 1.0 : (double) count / reqs;
+  }
+
+  double hitRate() {
+    return getRate(hitCount);
+  }
+
+  double missRate() {
+    return getRate(missCount);
```

Comment:
This is also used in the lambda expressions in the registerMetrics method. Without getRate, it would be very cumbersome in there. So I would keep this helper method.

Reply:
dropped the helper method